docs/docs/how-to/custom-configuration/es-modules.md

```diff
@@ -0,0 +1,160 @@
+---
+title: ES Modules (ESM) and Gatsby
+examples:
+  - label: Using MDX
+    href: "https://github.com/gatsbyjs/gatsby/tree/master/examples/using-mdx"
+---
+
+## Introduction
+
+The ECMAScript module (ESM) format is the [official TC39 standard](https://tc39.es/ecma262/#sec-modules) for packaging JavaScript. For many years, [CommonJS (CJS)](https://nodejs.org/api/modules.html#modules-commonjs-modules) was the de facto standard in Node.js. You can author [`gatsby-config`](/docs/reference/config-files/gatsby-config/) and [`gatsby-node`](/docs/reference/config-files/gatsby-node/) in ESM syntax.
+
+This feature was added in `gatsby@5.3.0`.
+
+## Prerequisites
+
+- A Gatsby project set up with `gatsby@5.3.0` or later. (Need help creating one? Follow the [Quick Start](/docs/quick-start/))
+
+## Usage in Gatsby
+
+Generally speaking you need to follow the official standard as explained in the [Node.js documentation](https://nodejs.org/api/esm.html).
+
+### `gatsby-config`
+
+Create a `gatsby-config.mjs` file. Here's an example `gatsby-config` using ESM syntax:
+
+```js:title=gatsby-config.mjs
+const config = {
+  siteMetadata: {
+    title: `ESM in Gatsby`,
+  },
+}
+
+export default config
+```
+
+### `gatsby-node`
+
+Create a `gatsby-node.mjs` file and use any of the [Node APIs](/docs/reference/config-files/gatsby-node/) as usual. Here's an example `gatsby-node` using ESM syntax:
+
+```js:title=gatsby-node.mjs
+export const onPostBuild = () => {
+  console.log("Build is done!")
+}
+```
+
+## Migrating from CommonJS to ES Modules
+
+- Use `import`/`export` syntax instead of `require`/`module.exports`
+- File extensions in imports are [mandatory](https://nodejs.org/api/esm.html#mandatory-file-extensions)
+- You can replicate the `__dirname` call with `import.meta.url`:
+
+  ```js
+  import { dirname } from "path"
+  import { fileURLToPath } from "url"
+
+  const __dirname = dirname(fileURLToPath(import.meta.url))
+  ```
+
+- You can replicate `require.resolve` with `createRequire`:
+
+  ```js
+  import { createRequire } from "module"
+
+  const require = createRequire(import.meta.url)
+  ```
+
+The documents [Interopability with CommonJS](https://nodejs.org/api/esm.html#interoperability-with-commonjs) and [Differences between ES Modules and CommonJS](https://nodejs.org/api/esm.html#differences-between-es-modules-and-commonjs) also apply to ESM in Gatsby.
+
+Here's how you'd migrate a `gatsby-config.js` file to `gatsby-config.mjs`.
+
+**Before:**
+
+```js:title=gatsby-config.js
+const { siteUrl } = require(`./defaults`)
+
+module.exports = {
+  siteMetadata: {
+    title: `Using CJS`,
+    siteUrl,
+  },
+  plugins: [
+    {
+      resolve: `gatsby-source-filesystem`,
+      options: {
+        name: `posts`,
+        path: `${__dirname}/content/posts/`,
+      },
+    },
+    {
+      resolve: `gatsby-source-filesystem`,
+      options: {
+        name: `images`,
+        path: require.resolve(`./content/images`),
```

Comment:
is this even valid? for this to work there would need to be `index.js` file in `./content/images`?

I think only place we really used `require.resolve` would be in `gatsby-node` when creating pages and figuring out path for `component` in `createPage`?

Reply:
Fixed in https://github.com/gatsbyjs/gatsby/pull/37934/commits/558aafbf32f912906457385cf2042a7d5bbfeb54